### PR TITLE
Add Pokemon#hasBeenActive

### PR DESCRIPTION
**Pull request by KrisXV**
Currently not used for anything, but it will be used for OMs in the future at the very least. Also potentially useful for future mechanics.

### CODE REVIEW

**Comment by Zarel** [comment]
Will it remain `0` if a Pokémon switches in and out the same turn? (e.g. an intentional switch followed by phazing?) Is that what you want?

**Comment by KrisXV** [comment]
Not sure if it does that, but yes that would be preferable because thats not truly an active turn

File: data/moves.js [hunk 1]
```diff
@@ -9824,6 +9824,7 @@ let BattleMovedex = {
 			}
 			// Instructed Fake Out etc. should fail
 			target.activeTurns++;
+			target.totalActiveTurns++;
```
Comment (Zarel) [inline data/moves.js]: This is a horrible hack I think?

Comment (KrisXV) [reply data/moves.js]: Why? if it counts towards increasing the active Pokemon's turn count, it should increase the overall turn count.

Comment (Zarel) [reply data/moves.js]: Yes, and increasing the active turns is a horrible hack

Comment (Zarel) [reply data/moves.js]: Piling hacks upon hacks is a bad idea, I think.

Comment (KrisXV) [reply data/moves.js]: I'll remove it for now then.

**Comment by Zarel** [comment]
I'm going to leave this open while I figure out how bad of a hack this actually is.

**Comment by KrisXV** [comment]
can you figure it out within the next 4 days? The next projected OMotM will likely need this to exist.

**Comment by Zarel** [comment]
Is a boolean `hasBeenActive` enough for the next OMotM?

**Comment by KrisXV** [comment]
Yes

File: sim/battle.ts [hunk 2]
```diff
@@ -1465,6 +1465,7 @@ export class Battle {
 					sideStaleness = sideStaleness === 'external' ? sideStaleness : pokemon.staleness;
 				}
 				pokemon.activeTurns++;
+				pokemon.hasBeenActive = true;
```
Comment (Zarel) [inline sim/battle.ts]: This is the wrong place to set a previouslyActive flag...

Comment (KrisXV) [reply sim/battle.ts]: well where would it be then?

Comment (Zarel) [reply sim/battle.ts]: The point of switch-out seems like a sane place.

Comment (Zarel) [reply sim/battle.ts]: Although depending on how the OM is designed, the point of move use might also work? Depends on the use-case.

Comment (Zarel) [reply sim/battle.ts]: The variable name should reflect which one it tracks, though.

Comment (Zarel) [reply sim/battle.ts]: `previouslySwitchedIn` vs `previouslyUsedMove`

Comment (scheibo) [reply sim/battle.ts]: Should this not be a field on `m: PokemonModData;`?

Comment (Zarel) [reply sim/battle.ts]: It's sufficiently general-purpose that I'm fine with it here.

Comment (KrisXV) [reply sim/battle.ts]: It could be, but it seems like something that could have actual use in the future

**Comment by Zarel** [comment]
After looking into this, it looks like it's a mess.

Anyway, `activeTurns` should probably be replaced with either one or two things, or at least renamed something to better reflect its status as a weird hack.

**Comment by KrisXV** [comment]
I did not make this pull request with the intent of fixing a hack and I do not plan on doing that in regards to this anywhere in the near future.

**Comment by Zarel** [comment]
You know you don't have to; you could have just ignored it and set `hasBeenActive` on switch-out...

**Comment by KrisXV** [comment]
The way it was handled initially ie totalActiveTurns was the most preferable option so i will just resort to using PokemonModData until a non-hacky way is found to handle this

**Comment by Zarel** [comment]
Can you stop being vague and just tell me how the mod works? Considering that it also works as a boolean, I'd be surprised if that was the best solution.

**Comment by Zarel** [comment]
@KrisXV I never expected you to fix the hack. I fixed it myself in 2d78044604f47284e5.

Feel free to base your new implementation on either the new `activeTurns` or the new `activeMoveActions`, or even tell me if you want a `previouslySwitchedIn` or `previouslyUsedMove` boolean - I can implement those, too.

**Comment by KrisXV** [comment]
I can't tell you how it works because there's no point in trying to quickly implement this if it doesn't win OMotM. If you really want to know, there is a thread for the OM, which is led by TI.

https://www.smogon.com/forums/threads/shared-power.3660877/

**Comment by Zarel** [comment]
I don't understand why this needs a "hasn't switched in before" counter?

**Comment by KrisXV** [comment]
The abilities don't apply to all Pokemon unless the Pokemon that originally has said ability has been active during the battle.

**Comment by Zarel** [comment]
This definitely sounds like `previouslySwitchedIn` is the correct metric here.

**Comment by KrisXV** [comment]
I'm still unsure on where to implement that; I'm unfamiliar with pokemon.ts

**Comment by Zarel** [comment]
It's relatively easy to implement if you don't get distracted by making multiple refactors to the battle engine.

**Comment by Zarel** [comment]
ba4fab47d2259f77daf6f279baf

There, I did this for you.

It will be `0` before your first switch-in, `1` after, `2` after your second switch in, etc.

**Pull request closed without merging**